boa_engine/src/builtins/intl/mod.rs

```diff
@@ -144,3 +147,540 @@ impl Intl {
         )))
     }
 }
+
+/// `MatcherRecord` type aggregates unicode `locale` string and unicode locale `extension`.
+///
+/// This is a return value for `lookup_matcher` and `best_fit_matcher` subroutines.
+#[derive(Debug)]
+struct MatcherRecord {
+    locale: JsString,
+    extension: JsString,
+}
+
+/// Returns the position of the first found unicode locale extension in a given string.
+///
+/// If no extensions found, return the length of requested locale
+fn get_leftmost_unicode_extension_pos(requested_locale: &str) -> usize {
+    let ext_sep = "-u-";
+    let src_locale = requested_locale.to_lowercase();
+    let pos = src_locale.find(ext_sep);
+    match pos {
+        Some(idx) => idx,
+        None => src_locale.len(),
+    }
+}
+
+/// Trims unciode locale extensions from a given string if any.
+///
+/// For example:
+///
+/// - `ja-Jpan-JP-u-ca-japanese-hc-h12` becomes `ja-Jpan-JP`
+/// - `fr-FR` becomes `fr-FR`
+fn trim_unicode_extensions(requested_locale: &str) -> JsString {
+    let trim_pos = get_leftmost_unicode_extension_pos(requested_locale);
+    JsString::new(&requested_locale[..trim_pos])
+}
```

Comment:
Would it be possible to replace this with the `icu` crate?
IIRC it has functionality to parse and manipulate locale strings.

Reply:
@jedel1043 as far, as I understood, it is possible to use locale parsing, like specified [here](https://docs.rs/icu/latest/icu/locid/struct.Locale.html#examples), and then serialize language identifier and extensions to string. I would like to clarify one thing here. My understanding is that icu crate integration is still in progress.

Would you recommend including icu crate (to replace locale extension related methods) into this PR?

Thanks.

Reply:
[Done](https://github.com/boa-dev/boa/pull/2036/files#diff-1dd4af22d3a6142a8060d056e9d0489d84d8a4d9c68a8f6e7b359b8d3ab35499R230). Please review.